### ISSUE TITLE
com.netflix.graphql.dgs.internal.DgsSchemaProvider#invokeDataFetcher should reuse DefaultParameterNameDiscoverer

### ISSUE THREAD

**Issue by lmy654** [Netflix/dgs-framework#158]
Currently, `DgsSchemaProvider#invokeDataFetcher ` creates a `DefaultParameterNameDiscoverer` instance on every invoking. A `DefaultParameterNameDiscoverer` instance has three `ParameterNameDiscoverer` instances, their `getParameterNames` and related methods are all thread-safe. So we should reuse the `DefaultParameterNameDiscoverer` instance as much as possible.
Especially for `LocalVariableTableParameterNameDiscoverer`, which need to inspect class file to get method parameter name. 
```java
@Nullable
public class LocalVariableTableParameterNameDiscoverer implements ParameterNameDiscoverer {

	private static final Log logger = LogFactory.getLog(LocalVariableTableParameterNameDiscoverer.class);

	// marker object for classes that do not have any debug info
	private static final Map<Executable, String[]> NO_DEBUG_INFO_MAP = Collections.emptyMap();

	// the cache uses a nested index (value is a map) to keep the top level cache relatively small in size
	private final Map<Class<?>, Map<Executable, String[]>> parameterNamesCache = new ConcurrentHashMap<>(32);


	@Override
	@Nullable
	public String[] getParameterNames(Method method) {
		Method originalMethod = BridgeMethodResolver.findBridgedMethod(method);
		return doGetParameterNames(originalMethod);
	}

	@Override
	@Nullable
	public String[] getParameterNames(Constructor<?> ctor) {
		return doGetParameterNames(ctor);
	}

	@Nullable
	private String[] doGetParameterNames(Executable executable) {
		Class<?> declaringClass = executable.getDeclaringClass();
		Map<Executable, String[]> map = this.parameterNamesCache.computeIfAbsent(declaringClass, this::inspectClass);
		return (map != NO_DEBUG_INFO_MAP ? map.get(executable) : null);
	}

	/**
	 * Inspects the target class.
	 * <p>Exceptions will be logged, and a marker map returned to indicate the
	 * lack of debug information.
	 */
	private Map<Executable, String[]> inspectClass(Class<?> clazz) {
		InputStream is = clazz.getResourceAsStream(ClassUtils.getClassFileName(clazz));
		if (is == null) {
			// We couldn't load the class file, which is not fatal as it
			// simply means this method of discovering parameter names won't work.
			if (logger.isDebugEnabled()) {
				logger.debug("Cannot find '.class' file for class [" + clazz +
						"] - unable to determine constructor/method parameter names");
			}
			return NO_DEBUG_INFO_MAP;
		}
		try {
			ClassReader classReader = new ClassReader(is);
			Map<Executable, String[]> map = new ConcurrentHashMap<>(32);
			classReader.accept(new ParameterNameDiscoveringVisitor(clazz, map), 0);
			return map;
		}
		catch (IOException ex) {
			if (logger.isDebugEnabled()) {
				logger.debug("Exception thrown while reading '.class' file for class [" + clazz +
						"] - unable to determine constructor/method parameter names", ex);
			}
		}
		catch (IllegalArgumentException ex) {
			if (logger.isDebugEnabled()) {
				logger.debug("ASM ClassReader failed to parse class file [" + clazz +
						"], probably due to a new Java class file version that isn't supported yet " +
						"- unable to determine constructor/method parameter names", ex);
			}
		}
		finally {
			try {
				is.close();
			}
			catch (IOException ex) {
				// ignore
			}
		}
		return NO_DEBUG_INFO_MAP;
	}
        // ... omit some code
}
```
Although it has `parameterNamesCache` for caching, but it won't work because of creating a `DefaultParameterNameDiscoverer` instance on every invoking. So every invoke comes in, it loads class file byte array, inspect them, and throw away, which make a huge stress on GC, and hurt the performance. We have clearly observed that on our test environment. We hava used Java Flight Recorder to profile memory allocation, and it showed:
![Very much byte array allocation](https://e3f49eaa46b57.cdn.sohucs.com/2021/3/16/19/11/MTAwMTAyXzE2MTU4OTMxMDQ1NjQ=.jpg)
So I recommend to reuse the `DefaultParameterNameDiscoverer` instance. `DgsSchemaProvider` can be modified like this:
```kotlin
package com.netflix.graphql.dgs.internal

// ... omit imports

class DgsSchemaProvider(
    private val applicationContext: ApplicationContext,
    private val federationResolver: Optional<DgsFederationResolver>,
    private val existingTypeDefinitionRegistry: Optional<TypeDefinitionRegistry>,
    private val mockProviders: Optional<Set<MockProvider>>
) {

    val dataFetcherInstrumentationEnabled = mutableMapOf<String, Boolean>()
    val entityFetchers = mutableMapOf<String, Pair<Any, Method>>()

    private val logger = LoggerFactory.getLogger(DgsSchemaProvider::class.java)

    private val objectMapper = jacksonObjectMapper().registerModule(JavaTimeModule())

    private val defaultParameterNameDiscoverer = DefaultParameterNameDiscoverer()

   // ... omit some code

   private fun invokeDataFetcher(method: Method, dgsComponent: Any, environment: DataFetchingEnvironment): Any? {
        return try {
            val args = mutableListOf<Any?>()
            val parameterNames = defaultParameterNameDiscoverer.getParameterNames(method) ?: emptyArray()
            method.parameters.forEachIndexed { idx, parameter ->

                // ... omit some code

            method.invoke(dgsComponent, *args.toTypedArray())
        } catch (ex: InvocationTargetException) {
            throw ex.targetException
        }
    }
    // ... omit some code
}
```

**Comment by paulbakker** [comment]
That's a really great find @lmy654! We only started using `ParameterNameDiscoverer` in the last release.

I'll work on a fix.

**Comment by paulbakker** [comment]
@lmy654 your suggested fix is in #160.
